jerry-ext/include/jerryscript-ext/module.h

```diff
@@ -45,15 +51,59 @@ typedef struct jerryx_native_module_t
  */
 #ifdef ENABLE_INIT_FINI
 #ifdef _MSC_VER
-#error "`FEATURE_INIT_FINI` build flag isn't supported on Windows, because Microsoft Visual C/C++ Compiler \
-doesn't support library constructors and destructors."
-#endif
-#define JERRYX_MODULE_CONSTRUCTOR_ATTRIBUTE __attribute__((constructor))
-#define JERRYX_MODULE_DESTRUCTOR_ATTRIBUTE __attribute__((destructor))
+
+typedef int (__cdecl *jerryx_msvc_section_function_t) (void);
+
+/* Only Visual Studio 2008 and Upper support for __pragma keyword */
+/* https://msdn.microsoft.com/en-us/library/d9x1s805(v=vs.90).aspx */
+#if _MSC_VER >= 1500
+#ifdef _M_X64
+#define JERRYX_MSVC_INCLUDE_SYM(s) comment(linker, "/include:" #s)
+#else
+#define JERRYX_MSVC_INCLUDE_SYM(s) comment(linker, "/include:_" #s)
+#endif /* _M_X64 */
+
+#define JERRYX_MSVC_DECL_SECTION_FUNCTION_POINTER(sec_name, func_name) \
+  __pragma(data_seg(push))                                             \
+      __pragma(section(sec_name, read)) __declspec(allocate(sec_name)) \
+          JERRYX_EXTERN_C jerryx_msvc_section_function_t               \
+              func_name##_section = func_name;                         \
+  __pragma(data_seg(pop))                                              \
+      __pragma(JERRYX_MSVC_INCLUDE_SYM(func_name##_section))
+
+#define JERRYX_MODULE_CONSTRUCTOR(construct)       \
+  static int __cdecl construct##_constructor(void) \
+  {                                                \
+    construct();                                   \
+    return 0;                                      \
+  }                                                \
+  JERRYX_MSVC_DECL_SECTION_FUNCTION_POINTER(".CRT$XCU", construct##_constructor)
+
+#define JERRYX_MODULE_DESTRUCTOR(destruct)       \
+  static int __cdecl destruct##_destructor(void) \
+  {                                              \
+    destruct();                                  \
+    return 0;                                    \
+  }                                              \
+  JERRYX_MSVC_DECL_SECTION_FUNCTION_POINTER(".CRT$XTU", destruct##_destructor)
```

Comment:
I'm not sure about these trampoline functions. Quite recently I had to work on code that gets auto-executed on load, and managed to get it work on Windows/MSVC, too, with a similar but simpler approach:

```c
#ifdef _MSC_VER
  #pragma section(".CRT$XCU", read)
  #define JERRYX_MODULE_CONSTRUCTOR2(f, prefix) \
    static void f (void); \
    __declspec(allocate(".CRT$XCU")) void (*f##_) (void) = f; \
    __pragma(comment(linker, "/include:" prefix #f "_")) \
    static void f (void)
  #ifdef _WIN64
    #define JERRYX_MODULE_CONSTRUCTOR(f) JERRYX_MODULE_CONSTRUCTOR2 (f, "")
  #else
    #define JERRYX_MODULE_CONSTRUCTOR(f) JERRYX_MODULE_CONSTRUCTOR2 (f, "_")
  #endif
#else
  #define JERRYX_MODULE_CONSTRUCTOR(f) \
    static void f (void) __attribute__((constructor)); \
    static void f (void)
#endif
```

Notes: Macro names have been adapted to the current scenario, but the pasted code is not feature-complete. E.g., it does not deal with destructors (but should be easy to adapt further). Static is also hard-coded. Etc.

Reply:
> I'm not sure about these trampoline functions. Quite recently I had to work on code that gets auto-executed on load, and managed to get it work on Windows/MSVC, too, with a similar but simpler approach:
> 
> ```c
> #ifdef _MSC_VER
>   #pragma section(".CRT$XCU", read)
>   #define JERRYX_MODULE_CONSTRUCTOR2(f, prefix) \
>     static void f (void); \
>     __declspec(allocate(".CRT$XCU")) void (*f##_) (void) = f; \
>     __pragma(comment(linker, "/include:" prefix #f "_")) \
>     static void f (void)
>   #ifdef _WIN64
>     #define JERRYX_MODULE_CONSTRUCTOR(f) JERRYX_MODULE_CONSTRUCTOR2 (f, "")
>   #else
>     #define JERRYX_MODULE_CONSTRUCTOR(f) JERRYX_MODULE_CONSTRUCTOR2 (f, "_")
>   #endif
> #else
>   #define JERRYX_MODULE_CONSTRUCTOR(f) \
>     static void f (void) __attribute__((constructor)); \
>     static void f (void)
> #endif
> ```
> 
> Notes: Macro names have been adapted to the current scenario, but the pasted code is not feature-complete. E.g., it does not deal with destructors (but should be easy to adapt further). Static is also hard-coded. Etc.

I've seen you code, but have mistake in this code, The #pragma section(".CRT$XCU", read) should not used unconditionally, should also convert to use __param(section(".CRT$XCU", read))
and also it's didn't consider the amd64 and x86 differencies in
```
    __pragma(comment(linker, "/include:" prefix #f "_")) \

```

Reply:
The #pragma simply defines a section. What problem does it cause if it is left empty? But it can be converted to __pragma().

As for the amd64/x86 differences: if you mean the underscore prefix necessary on x86, it is there. If you mean something else, please, specify.

And, as mentioned below the code snippet, this is not a fully working suggestion but an example only that may need further adaptation. Feel free to adapt. The key point of the comment was to get rid of the trampolines.

Reply:
I am talking about  #pragma section(".CRT$XCU", read), without   __pragma (data_seg (push))  
 __pragma (data_seg (pop))

It's will affect other section definition.

The  underscore prefix necessary on x86, but should not on x64 that's why macro JERRYX_MSVC_INCLUDE_SYM defined

Reply:
As Igor says, #ifdef is a compile-time check.  But if you really do want a compile-time check, you don’t want to use _WIN64 because that is not not automatically defined when setting up the project configuration.  Instead, use _M_AMD64.  http://msdn.microsoft.com/en-us/library/b0084kay.aspx

Reply:
There is only a single variable declaration/initialization between the segment push/pop pairs, and that variable has an explicit allocation declspec. Please, explain in detail why the push/pop is needed then.

I'm fine with _M_AMD64, but as I recall from the recent jerry-math PR, we have also used _M_X64 as well.

(And who is Igor? I'm losing context)

Reply:
Ignore Igor, copied from blog, I am not sure if we need use _WIN64 or _M_AMD64, I don't know what's the C ABI of ARM for msvc and didn't find the document for it.
But the push/pop are necessary.

#ifdef _MSC_VER
  #pragma section(".CRT$XCU", read)
Means the following code are setting ".CRT$XCU" to read

but after this, anyone can modify the ".CRT$XCU" attribute.
that's why `__pragma (section (sec_name, read))` needed.
As between the module.h and the code you define the constructor can have any kinds of code.

the pop and push is a standard way to guard the __pragma (section) not pollute other code. because in C++
You won't know where the macro are called.
And what's kinds of code after the macro

Reply:
> Ignore Igor, copied from blog

How about not copying it then...

> I am not sure if we need use _WIN64 or _M_AMD64, I don't know what's the C ABI of ARM for msvc and didn't find the document for it.

I was not referring to ARM at all. I was suggesting to use `#if defined (_M_AMD64) || defined (_M_X64)` (just like it is done at https://github.com/jerryscript-project/jerryscript/blob/master/jerry-math/jerry-math-internal.h#L41 )

> But the push/pop are necessary.
>
> #ifdef _MSC_VER
> #pragma section(".CRT$XCU", read)
> Means the following code are setting ".CRT$XCU" to read
>
> but after this, anyone can modify the ".CRT$XCU" attribute.
> that's why __pragma (section (sec_name, read)) needed.
> As between the module.h and the code you define the constructor can have any kinds of code.
> 
> the pop and push is a standard way to guard the __pragma (section) not pollute other code. 

Unfortunately, I'm not sure I can follow your argument.

The `section` pragma does _not_ change where following code or data is placed. See: "Creates a section in an OBJ file. [...] Once a section is defined, it remains valid for the rest of the compilation. However, you must use __declspec(allocate), or nothing is placed in the section." (https://docs.microsoft.com/en-us/cpp/preprocessor/section?view=msvc-160)

This means that in the code snippet below, between the push and pop, there is a section definition that does not affect how following data or code is placed, and a variable declaration that is explicitly allocated in a given section.

```c
  __pragma (data_seg (push))                                              \
      __pragma (section (sec_name, read)) __declspec(allocate (sec_name)) \
          JERRYX_MSCV_EXTERN_C jerryx_msvc_section_function_t             \
              func_name##_section = func_name;                            \
  __pragma (data_seg (pop))                                               \
```

 According to my understanding this makes the push/pop useless. Please, explain where I'm wrong.

> because in C You won't know where the macro are called.

I think I have some basic knowledge of C, but thanks for the heads up.

Reply:
Ah, sorry for that, you mean  __pragma (section (sec_name, read)) only take effect on `func_name##_section`, that's make sense
I forgot that may code was compatiable to MSVC 2005 and below but now it's unnecessary anymore. I'll update the patch.

Thanks for the review

Reply:
This is the MSVC demo:
```
#define CATCH_ALL_DUAL \
CATCH(COleException, e) \
{ \
_hr = e->m_sc; \
} \
AND_CATCH_ALL(e) \
{ \
__pragma(warning(push)) \
__pragma(warning(disable:6246)) /*disable _ctlState prefast warning*/ \
AFX_MANAGE_STATE(pThis->m_pModuleState); \
__pragma(warning(pop)) \
_hr = DualHandleException(_riidSource, e); \
} \
END_CATCH_ALL \
return _hr; \
```
Please taka a look at https://docs.microsoft.com/en-us/cpp/preprocessor/pragma-directives-and-the-pragma-keyword?view=msvc-160

Reply:
I'm sorry, but I don't understand why you are copying & linking this here. There was no discussion about warnings here, nor about exceptions.

Reply:
> I'm sorry, but I don't understand why you are copying & linking this here. There was no discussion about warnings here, nor about exceptions.

I am talking about __pragma(warning(push)) , __pragma(warning(pop)) 
For section it's the same.
Please note, it's msvc extention __pragma, not C99 _Pragma

Reply:
I totally understand how a push/pop mechanism works: for warnings, for data section, or -- say -- for machine stack. What I don't understand is its relevance here. I've given a detailed explanation why I think that section pragma does not have to be guarded by a data_seg push/pops (because section pragma does not change where following data is placed, it only creates a section for future reference by allocate declspec).

For warnings, it is pretty much different, because a warning(disable:N) indeed has effect on what follows. That's why push/pop guards are needed, to save/restore state.

It seemed that you agreed because there was a comment about that -- which is now gone, most probably deleted. I've recovered it from my mail archive:

> Ah, sorry for that, you mean __pragma (section (sec_name, read)) only take effect on func_name##_section, that's make sense
> I forgot that may code was compatiable to MSVC 2005 and below but now it's unnecessary anymore. I'll update the patch.
>
> Thanks for the review

I don't think that erasing / rewriting history is a good approach and does not help trust.

Plus, you are communicating in half sentences and/or without context:

> I am talking about __pragma(warning(push)) , __pragma(warning(pop))

No, you were not talking about that, you just copy-pasted a code snippet and a link without any explanation why you though it would be relevant. I'm not good at reading minds -- and I'm not eager to learn it either.

I think I'll suspend communication and review until attitude changes for the better.